File: day11/day11.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stack>

using namespace std;

void part1(vector<vector<int>> energyLevel);
void part2(vector<vector<int>> energyLevel);
vector<vector<int>> readInput();
int step(vector<vector<int>> &energyLevel);
int nStep(vector<vector<int>> &energyLevel, int n);
int getSynchronousFlash(vector<vector<int>> &energyLevel);


int main() {
    vector<vector<int>> energyLevel = readInput();
    part1(energyLevel);
    part2(energyLevel);
    return 0;
}

void part1(vector<vector<int>> energyLevel) {
    int totalFlashCount = nStep(energyLevel, 100);
    cout << "Part 1: " << totalFlashCount << endl;
}

void part2(vector<vector<int>> energyLevel) {
    int count = getSynchronousFlash(energyLevel);
    cout << "Part 2: " << count << endl;
}

vector<vector<int>> readInput() {
    vector<vector<int>> energyLevel;
    string line;
    while (cin >> line) {
        vector<int> row;
        for (char number : line) {
            int digit = number - '0';
            row.push_back(digit);
        }
        energyLevel.push_back(row);
    }

    return energyLevel;
}

int step(vector<vector<int>> &energyLevel) {
    stack<pair<int, int>> canFlash;
    int flashCount = 0;
    // Increment by 1
    for (int i = 0; i < energyLevel.size(); i++) {
        for (int j = 0; j < energyLevel[0].size(); j++) {
            energyLevel[i][j]++;
            if (energyLevel[i][j] == 10) {
                canFlash.push({i, j});
            }
        }
    }
    // Flash coordinates and carry out chain reactions
    while (!canFlash.empty()) {
        pair<int, int> coordinate = canFlash.top();
        canFlash.pop();
        int x = coordinate.first, y = coordinate.second;
        energyLevel[x][y] = 11;
        flashCount++;
        vector<pair<int, int>> surroundingCoordinates = {{x - 1, y - 1}, {x - 1, y}, {x - 1, y + 1},
                                                         {x, y - 1}, {x, y}, {x, y + 1},
                                                         {x + 1, y - 1}, {x + 1, y}, {x + 1, y + 1}};
        for (auto surroundingCoordinate : surroundingCoordinates) {
            int x1 = surroundingCoordinate.first, y1 = surroundingCoordinate.second;
            if (x1 < 0 || x1 >= energyLevel.size() || y1 < 0 || y1 >= energyLevel[0].size()) {
                continue;
            }
            if (energyLevel[x1][y1] >= 10) {
                continue;
            }
            energyLevel[x1][y1]++;
            if (energyLevel[x1][y1] == 10) {
                canFlash.push({x1, y1});
            }
        }
    }
    // Reset coordinates that have flashed back to 0
    for (int i = 0; i < energyLevel.size(); i++) {
        for (int j = 0; j < energyLevel[0].size(); j++) {
            if (energyLevel[i][j] == 11) {
                energyLevel[i][j] = 0;
            }
        }
    }
    return flashCount;
}

int nStep(vector<vector<int>> &energyLevel, int n) {
    int totalFlashCount = 0;
    for (int i = 0; i < n; i++) {
        totalFlashCount += step(energyLevel);
    }
    return totalFlashCount;
}

int getSynchronousFlash(vector<vector<int>> &energyLevel) {
    int count = 0, flashCount = 0, maxFlash = energyLevel.size() * energyLevel[0].size();
    while (flashCount != maxFlash) {
        flashCount = step(energyLevel);
        count++;
    }
    return count;
}

File: day11/day11.js
const fs = require('fs');

let input = fs.readFileSync('./input.in').toString().split(/\r?\n/);

if (input[input.length - 1] === '') {
    input.pop();
}

input = input.map(row => row.split('').map(x => parseInt(x)));

// Perform 1 step
doStep = input => {
    let canFlash = [];  // Stores all the octopi that needs to be flashed
    let count = 0;  // Count number of flashes

    // Increment all by 1
    input = input.map(row => row.map(x => x += 1));
    for (let i = 0; i < input.length; i++) {
        for (let j = 0; j < input[0].length; j++) {
            if (input[i][j] === 10) {
                // Means this octopus is ready to flash
                canFlash.push([i, j]);
            }
        }
    }

    while (canFlash && canFlash.length) {
        [i, j] = canFlash.pop();
        input[i][j] = 11;
        count++;
        let surr = [[i - 1, j - 1], [i - 1, j], [i - 1, j + 1], 
                    [i, j - 1], [i, j + 1], 
                    [i + 1, j - 1], [i + 1, j], [i + 1, j + 1]]
        for (coor of surr) {
            if (coor[0] < 0 || coor[0] >= input.length || coor[1] < 0 || coor[1] >= input.length) {
                // Octopus isn't in the grid, skip it
                continue;
            }
            if (input[coor[0]][coor[1]] >= 10) {
                // The octopus has already flashed, skip it
                continue;
            }
            input[coor[0]][coor[1]]++;
            if (input[coor[0]][coor[1]] == 10) {
                // The adjacent flash caused this octopus to be ready to flash
                canFlash.push(coor);
            }
        }
    }

    // Reset all octopi that have flashed
    for (let i = 0; i < input.length; i++) {
        for (let j = 0; j < input[0].length; j++) {
            if (input[i][j] == 11) {
                input[i][j] = 0;
            }
        }
    }

    return [input, count];

}

part1 = input => {
    let totSteps = 100;
    let totFlashes = 0;
    for (let i = 0; i < totSteps; i++) {
        [input, count] = doStep(input);
        totFlashes += count;
    }
    console.log("Part 1 Answer: %i", totFlashes);
}

part2 = input => {
    let count = 0;
    let flashCount = 0;
    while (flashCount !== input.length * input[0].length) {
        [input, flashCount] = doStep(input);
        count++;
    }
    console.log("Part 2 Answer: %i", count);
}

part1(input);
part2(input);
